### PR TITLE
Glean events for subscription management and account deletion

### LINKED ISSUE #3627: Instrumentation subscription management (client)
Add glean events for:

* Instrument account deletion
* Viewing the subscription management page
* Clicking on the "manage account" button (if there is one)
* Unhandled subscription plan interval

There are tickets related to this in the tips and tricks epic.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/VPN-2291)
┆author: Sarah Bird

### PR DESCRIPTION
**Pull request by flozia**
## Description

Adds the requested Glean events for the subscription management and account deletion flow. Needs data review before merging.

## Reference

Resolves #3627

## Checklist
    
- [x] My code follows the style guidelines for this project
- [x] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [x] I have performed a self review of my own code
- [ ] I have commented my code PARTICULARLY in hard to understand areas
- [ ] I have added thorough tests where needed

### CODE REVIEW

**Comment by flozia** [comment]
Had to close [PR #3722](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/3722) in favour of this PR due to unresolvable diffs with main caused by changed commit hashes.

File: glean/metrics.yaml [hunk 1]
```diff
@@ -332,6 +332,55 @@ sample:
...
+    expires: never
+    extra_keys:
+      state:
+        description: |
+          The name of the authentication state, e.g. "sign-in", "sign-out",
+          "totp", etc. The list of states can be found in
+          `authenticationinapp.h`.
+        type: string
+
+  auth_subscription_mgmt_step:
```
Comment (birdsarah) [inline glean/metrics.yaml:359]: I'd like to mirror `authentication_inapp_step`. My suggestion is `authentication_sub_manage_step`.

File: glean/metrics.yaml [hunk 2]
```diff
@@ -332,6 +332,55 @@ sample:
           `authenticationinapp.h`.
         type: string
 
+  auth_account_deletion_step:
```
Comment (birdsarah) [inline glean/metrics.yaml:335]: I'd like to mirror `authentication_inapp_step`. My suggestion is `authentication_account_delete_step`.

File: glean/metrics.yaml (hunk 1 above)
Comment (flozia) [reply glean/metrics.yaml]: Updated the name.

File: glean/metrics.yaml (hunk 2 above)
Comment (flozia) [reply glean/metrics.yaml]: Unfortunately `authentication_account_delete_step` is too long and causes the unit test to throw. The max length seems to be 30 chars.
```
/home/runner/work/mozilla-vpn-client/mozilla-vpn-client/glean/metrics.yaml:
ERROR running glean_parser v6.1.1
    ```
    sample: authentication_account_del_step
    ```
    'authentication_account_del_step' is not valid under any of the given
    schemas
        'authentication_account_del_step' is too long
```

File: src/authenticationinapp/authenticationinapp.cpp [hunk 3]
```diff
@@ -48,9 +48,26 @@ void AuthenticationInApp::setState(State state) {
   m_state = state;
   emit stateChanged();
 
-  emit MozillaVPN::instance()->recordGleanEventWithExtraKeys(
-      GleanSample::authenticationInappStep,
-      {{"state", QVariant::fromValue(state).toString()}});
+  if (m_session &&
```
Comment (bakulf) [inline src/authenticationinapp/authenticationinapp.cpp:51]: In this way, if `m_session` is null, we are unable to send the glean event. What about if we change `setState` like this:
```
void AuthenticationInApp::setState(State state,  AuthenticationInAppSession* session) {
...
```

And in the session destroy slot, you do:
```
AuthenticationInAppSession* session = m_session;
m_session = nullptr;
setState(StateInitializing, session);
```

**Review by bakulf — CHANGES_REQUESTED** [review]

**Comment by flozia** [comment]
Data review is requested and can be tracked under [Bug 1774568](https://bugzilla.mozilla.org/show_bug.cgi?id=1774568).

**Comment by flozia** [comment]
We received the result `datareview+` for the data review mentioned in the previous comment.

File: glean/metrics.yaml [hunk 4]
```diff
@@ -332,6 +332,55 @@ sample:
...
+    type: event
+    lifetime: ping
+    send_in_pings:
+      - main
+    description: |
+      The user has advanced in the authentication flow for the account deletion.
+    bugs:
+      - https://github.com/mozilla-mobile/mozilla-vpn-client/issues/2956
+    data_reviews:
+      - TBD
```
Comment (birdsarah) [inline glean/metrics.yaml:345]: Please fill in the data review bug number

**Review by birdsarah — APPROVED** [review]

File: src/authenticationinapp/authenticationinapp.cpp [hunk 5]
```diff
@@ -44,13 +44,30 @@ AuthenticationInApp::~AuthenticationInApp() {
...
-void AuthenticationInApp::setState(State state) {
+void AuthenticationInApp::setState(State state,
+                                   AuthenticationInAppSession* session) {
   m_state = state;
   emit stateChanged();
 
-  emit MozillaVPN::instance()->recordGleanEventWithExtraKeys(
-      GleanSample::authenticationInappStep,
-      {{"state", QVariant::fromValue(state).toString()}});
+  if (session && session->type() == AuthenticationInAppSession::TypeDefault) {
```
Comment (bakulf) [inline src/authenticationinapp/authenticationinapp.cpp:52]: what about a `switch` ?
if (session) {
  char* gleanSample = nullptr;
  switch (session->type()) {
  case  AuthenticationInAppSession::TypeDefault: gleanSample = GleanSample::authenticationInappStep; break;
  ...
  default:
   logger.error() << "Glean samples and Auth-in-app session types are out of sync";
    }

 Q_ASSERT(gleanSample);
emit MozillaVPN::instance()->recordGleanEventWithExtraKeys(gleanSample, {{"state", QVariant::fromValue(state).toString()}});
```

**Review by bakulf — CHANGES_REQUESTED** [review]

**Review by bakulf — APPROVED** [review]

File: src/authenticationinapp/authenticationinapp.cpp [hunk 6]
```diff
@@ -44,13 +44,34 @@ AuthenticationInApp::~AuthenticationInApp() {
...
 }
 
-void AuthenticationInApp::setState(State state) {
+void AuthenticationInApp::setState(State state,
+                                   AuthenticationInAppSession* session) {
   m_state = state;
   emit stateChanged();
 
+  Q_ASSERT(session);
+  const char* gleanSample = nullptr;
```
[the remark this answers is not in the document]
Comment (flozia) [reply src/authenticationinapp/authenticationinapp.cpp:53]: As discussed: `const` here is fine because the value of the pointed element is immutable.

**Pull request merged**